renderer/backend/DrawBatchExecutor.cpp

```diff
@@ -0,0 +1,167 @@
+	cmd.count = surf->numIndexes;
+	cmd.instanceCount = 1;
+	cmd.firstIndex = surf->indexCache.offset / sizeof(glIndex_t);
+	cmd.baseVertex = surf->shadowCache.offset / sizeof(shadowCache_t);
+	cmd.baseInstance = idx;
+
+	if (surf->frontendGeo)
+		numVerts += surf->frontendGeo->numVerts;
+	numIndexes += surf->numIndexes;
+	shadows = true;
+}
+
+void DrawBatchExecutor::DrawBatch() {
+	int numDrawCalls = currentIndex;
+
+	if (numDrawCalls == 0) {
+		currentIndex = 0;
+		currentCommands = nullptr;
+		maxDrawCommands = 0;
+		return;		
+	}
+
+	vertexCache.BindIndex();
+
+	if (ShouldUseMultiDraw()) {
+		if (!drawIdVertexEnabled) {
+			drawIdVertexEnabled = true;
+			qglEnableVertexAttribArray( Attributes::Default::DrawId );
+		}
+		drawCommandBuffer.Commit( reinterpret_cast< byte* >( currentCommands ), numDrawCalls * sizeof(DrawElementsIndirectCommand) );
+		qglMultiDrawElementsIndirect(GL_TRIANGLES, GL_INDEX_TYPE, drawCommandBuffer.BufferOffset(currentCommands), numDrawCalls, 0);
+	} else {
+		if (drawIdVertexEnabled) {
+			drawIdVertexEnabled = false;
+			qglDisableVertexAttribArray( Attributes::Default::DrawId );
+		}
+		for (int i = 0; i < numDrawCalls; ++i) {
+			DrawElementsIndirectCommand &cmd = currentCommands[i];
+			const void *indexOffset = reinterpret_cast< const void* >( cmd.firstIndex * sizeof(glIndex_t) );
+			qglVertexAttribI1i(Attributes::Default::DrawId, i);
+			qglDrawElementsBaseVertex(GL_TRIANGLES, cmd.count, GL_INDEX_TYPE, indexOffset, cmd.baseVertex);
+		}
+	}
+
+	currentIndex = 0;
+	currentCommands = nullptr;
+	maxDrawCommands = 0;
+
+	if ( r_showPrimitives.GetBool() && !backEnd.viewDef->IsLightGem() && backEnd.viewDef->viewEntitys ) {
+		if (shadows) {
+			backEnd.pc.c_shadowElements += numDrawCalls;
+			backEnd.pc.c_shadowIndexes += numIndexes;
+			backEnd.pc.c_shadowVertexes += numVerts;
+		} else {
+			backEnd.pc.c_drawElements += numDrawCalls;
+			backEnd.pc.c_drawIndexes += numIndexes;
+			backEnd.pc.c_drawVertexes += numVerts;
+			backEnd.pc.c_vboIndexes += numIndexes;
+		}
+	}
+}
+
+void DrawBatchExecutor::Lock() {
+	drawCommandBuffer.Lock();
+}
+
+bool DrawBatchExecutor::ShouldUseMultiDraw() const {
+	return GLAD_GL_ARB_multi_draw_indirect && r_useMultiDrawIndirect;
+}
+
+void DrawBatchExecutor::InitDrawIdBuffer() {
+	qglGenBuffers(1, &drawIdBuffer);
+	qglBindBuffer(GL_ARRAY_BUFFER, drawIdBuffer);
+	std::vector<uint32_t> drawIds (MAX_DRAW_COMMANDS);
+	for (uint32_t i = 0; i < MAX_DRAW_COMMANDS; ++i) {
+		drawIds[i] = i;
+	}
+	qglBufferData(GL_ARRAY_BUFFER, drawIds.size(), drawIds.data(), GL_STATIC_DRAW);
+	qglBindBuffer(GL_ARRAY_BUFFER, 0);
+
+	qglVertexAttribIFormat( Attributes::Default::DrawId, 1, GL_UNSIGNED_INT, 0 );
+	qglBindVertexBuffer( Attributes::Default::DrawId, drawIdBuffer, 0, sizeof(uint32_t) );
+	qglVertexAttribBinding( Attributes::Default::DrawId, Attributes::Default::DrawId );
+	qglVertexBindingDivisor( Attributes::Default::DrawId, 1 );
+	qglEnableVertexAttribArray( Attributes::Default::DrawId );
```

Comment:
What an abomination!
Just wanted to clarify: is it really necessary here?

Reply:
What exactly do you mean? The setup is necessary unfortunately, though I suppose I could move it to the RenderBackend. Don't really have any strong opinions on that :)